libraries/chain/db_market.cpp

```diff
@@ -859,38 +867,52 @@ bool database::fill_call_order( const call_order_object& order, const asset& pay
 
+   // calculate any margin call fees
+   asset margin_fee(0, pays.asset_id);
+   if (is_margin_call)
+   {
+      margin_fee = calculate_margin_fee( mia, pays );
+      FC_ASSERT( margin_fee.asset_id == pays.asset_id ); // margin fee should be paid in the debt asset
+      // margin fee should never be more than what the call order is receiving (but both could be zero)    
+      FC_ASSERT( margin_fee.amount == 0 || margin_fee.amount < pays.amount ); 
+   }
+
    optional<asset> collateral_freed;
-   modify( order, [&]( call_order_object& o ){
-            o.debt       -= receives.amount;
-            o.collateral -= pays.amount;
-            if( o.debt == 0 )
-            {
-              collateral_freed = o.get_collateral();
-              o.collateral = 0;
-            }
-            else
+   // adjust the order
+   modify( order, [&]( call_order_object& o ) {
+         o.debt       -= receives.amount;
+         o.collateral -= pays.amount;
+         if( o.debt == 0 ) // is the whole debt paid?
+         {
+            collateral_freed = o.get_collateral();
+            o.collateral = 0;
+         }
+         else // the debt was not completely paid
+         {
+            auto maint_time = get_dynamic_global_properties().next_maintenance_time;
+            // update call_price after core-343 hard fork,
+            // but don't update call_price after core-1270 hard fork
+            if( maint_time <= HARDFORK_CORE_1270_TIME && maint_time > HARDFORK_CORE_343_TIME )
             {
-               auto maint_time = get_dynamic_global_properties().next_maintenance_time;
-               // update call_price after core-343 hard fork,
-               // but don't update call_price after core-1270 hard fork
-               if( maint_time <= HARDFORK_CORE_1270_TIME && maint_time > HARDFORK_CORE_343_TIME )
-               {
-                  o.call_price = price::call_price( o.get_debt(), o.get_collateral(),
-                                                    mia.bitasset_data(*this).current_feed.maintenance_collateral_ratio );
-               }
+               o.call_price = price::call_price( o.get_debt(), o.get_collateral(),
+                     mia.bitasset_data(*this).current_feed.maintenance_collateral_ratio );
             }
+         }
       });
 
+   // distribute the margin fee
+   if (margin_fee.amount > 0)
+      distribute_market_fees( &order.borrower(*this), margin_fee.asset_id(*this), margin_fee );
+
    // update current supply
    const asset_dynamic_data_object& mia_ddo = mia.dynamic_asset_data_id(*this);
-
    modify( mia_ddo, [&receives]( asset_dynamic_data_object& ao ){
          ao.current_supply -= receives.amount;
       });
 
-   // Adjust balance
+   // If the whole debt is paid, adjust borrower's collateral balance
    if( collateral_freed.valid() )
-      adjust_balance( order.borrower, *collateral_freed );
+      adjust_balance( order.borrower, *collateral_freed - margin_fee );
```

Comment:
I think this change is wrong. The fee should be deducted from the collateral paying to the counterpart, but not from the remaining collateral.